src/operator/controllers/aws_iam/serviceaccount_old/serviceaccount_controller.go

```diff
@@ -0,0 +1,197 @@
+			if controllerutil.RemoveFinalizer(updatedServiceAccount, metadata.AWSRoleFinalizer) {
+				err := r.Client.Patch(ctx, updatedServiceAccount, client.MergeFrom(&serviceAccount))
+				if err != nil {
+					if apierrors.IsConflict(err) {
+						return ctrl.Result{Requeue: true}, nil
+					}
+					return ctrl.Result{}, errors.Wrap(err)
+				}
+			}
+
+		}
+		return ctrl.Result{}, nil
+	}
+
+	updatedServiceAccount := serviceAccount.DeepCopy()
+	if controllerutil.AddFinalizer(updatedServiceAccount, metadata.AWSRoleFinalizer) {
+		err := r.Client.Patch(ctx, updatedServiceAccount, client.MergeFrom(&serviceAccount))
+		if err != nil {
+			if apierrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, errors.Wrap(err)
+		}
+	}
+
+	if viper.GetBool(operatorconfig.EnableAWSRolesAnywhereKey) {
+		// In RolesAnywhere mode, the SPIFFE pod webhook, and not the reconciler, handles the role creation
+		return ctrl.Result{}, nil
+	}
+
+	shouldUpdateAnnotation, role, err := r.reconcileAWSRole(ctx, &serviceAccount)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err)
+	}
+
+	if shouldUpdateAnnotation {
+		if updatedServiceAccount.Annotations == nil {
+			updatedServiceAccount.Annotations = make(map[string]string)
+		}
+
+		updatedServiceAccount.Annotations[metadata.ServiceAccountAWSRoleARNAnnotation] = *role.Arn
+		err = r.Client.Patch(ctx, updatedServiceAccount, client.MergeFrom(&serviceAccount))
+		if err != nil {
+			if apierrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, errors.Wrap(err)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcileAWSRole ensures the AWS role for the service account exists
+// return values:
+// - modified: if AWS state was modified (and the ServiceAccount should be re-queued)
+// - role: the AWS IAM role
+// - err
+func (r *ServiceAccountReconciler) reconcileAWSRole(ctx context.Context, serviceAccount *corev1.ServiceAccount) (updateAnnotation bool, role *types.Role, err error) {
+	logger := logrus.WithFields(logrus.Fields{"serviceAccount": serviceAccount.Name, "namespace": serviceAccount.Namespace})
+	roleARN, ok := hasAWSAnnotation(serviceAccount)
+
+	// calling create in any case because this way we validate it is not soft-deleted and it is configured with the correct soft-delete strategy
+	role, err = r.awsAgent.CreateOtterizeIAMRole(ctx, serviceAccount.Namespace, serviceAccount.Name, r.shouldUseSoftDeleteStrategy(serviceAccount))
+	if err != nil {
+		return false, nil, fmt.Errorf("failed creating AWS role for service account: %w", err)
+	}
+	logger.WithField("arn", *role.Arn).Info("created AWS role for ServiceAccount")
+
+	// update annotation if it doesn't exist or if it is misconfigured
+	shouldUpdate := !ok || roleARN != *role.Arn
+
+	return shouldUpdate, role, nil
+}
+
+func (r *ServiceAccountReconciler) shouldUseSoftDeleteStrategy(serviceAccount *corev1.ServiceAccount) bool {
+	if r.markRolesAsUnusedInsteadOfDelete {
```

Comment:
I suggest changing the order of the "if global UseSoftDelete" and the "if per pod useSoft" statements here. This way, when both configurations are present, the "per pod" configuration will be given priority over the global one. It makes more sense. 
(I know it's my code 😅)